### PR TITLE
Tree API Discussion

### PR DESCRIPTION
**Pull request by DLehenbauer**
Creating a draft PR of Craig's branch so we can use GitHub's review tools to collaborate.

FYI - @CraigMacomber, @taylorsw04, @ruiterr, @PaulKwMicrosoft

### CODE REVIEW

File: experimental/dds/tree/src/interface-prototype/Anchors.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import { ConstraintEffect } from '../default-edits';
```
Comment (DLehenbauer) [inline experimental/dds/tree/src/interface-prototype/Anchors.ts:1]: From our discussion of ops, I was expecting Anchor to be something like:
```ts
interface Anchor {
    node: NodeId;
    trait: string;
    position: "first" | "before" | "after" | "last"
}
```

Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Anchors.ts:1]: I'll do some renames and add some comments to try and clarify this.

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	PrefetchFilter,
+	root,
+	SharedTree,
+	TreeNodeViewReadonly,
+} from './Checkout';
+import { Place, TreeNode } from './MutableAnchors';
+
+//////////////// Command examples //////////////
+
+// Inserts a node with the specified Definition and identifier at the specified Place, and return it.
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:23]: Minor typo: probably should say "returns".

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+export const insertExample = {
+	id: 'f73f004c-3b3e-42fe-b7c9-2e5e8793ca45' as CommandId,
+	run: (
+		context: CommandContext,
+		{ definition: def, identifier: id }: { definition: StableId; identifier: StableId },
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:28]: How about just "id" as the property name. I'm not even sure "identifier" is the right term, anyway.

Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:28]: And (especially if we make that "id" change above), let's just write { definition, id } - I'm not sure there's any benefit to abbreviating definition, and it just makes the method signature look bulky.

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		{ place }: { place: Place }
+	): TreeNode => {
+		const definition = context.loadDefinition(def);
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:31]: Feels very odd (and tedious) having to load a definition or an id in a context. I would have thought those were invariant across contexts.

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+		const identifier = context.loadNodeId(id);
+		const range = place.insert({ definition, identifier, traits: {} });
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:33]: Also, why would one be supplying the identifier? I would have thought SharedTree assigns identities, not the client.

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+		return range.start.adjacentNode(Side.After);
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:34]: range.start.adjacentNode(Side.After) seems a bit wordy. How about we add range.first and range.last? (If only we could implicitly coerce a single-node range to a TreeNode. :-/ )

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+			insertExample,
+			{ definition: context.stabilize(baz), identifier: newNodeId() },
+			{ place: a.childrenFromTrait(testTrait).start }
+		);
+	},
+};
+
+const filterAll = () => true;
+
+const fetchAll: PrefetchFilter = {
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:61]: Why don’t we put fetchAll in the API? It seems like it’ll be pretty common. Alternately, we could just make that the default behind-the-scenes for an optional second parameter to tree.checkout and any other methods one would be likely to pass this to.

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	children: filterAll,
+	traitChunks: filterAll,
+};
+
+const commands: CommandRegistry = [insertExample, doubleInsertExample];
+
+// Some dummy schema related data.
+const foo: Definition = 'Foo' as Definition;
+const bar: Definition = 'Bar' as Definition;
+const baz: Definition = 'Bar' as Definition;
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:72]: Typo: 'Bar' not 'Baz'

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+const testTrait: TraitLabel = 'testTrait' as TraitLabel;
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:74]: Given the 'as', can drop the type on the const. I think our linter would complain about the redundancy.

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+export async function exampleApp(tree: SharedTree): Promise<void> {
+	// Perform a full checkout of all data.
+	const checkout = await tree.checkout(commands, fetchAll);
+
+	// Example app policy: this app just watches the tree, and adds bar(baz) subtree under any new `foo` node.
+	checkout.on('viewChange', (before, after) => {
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:81]: Is this a common pattern, having a single method for registering change handlers? I was expecting either separate ".onViewChange = " registrations or for us to pass an interface with all the handlers, i.e. using property names instead of strings to specify the changes being handled.

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+const fetchAll: PrefetchFilter = {
+	value: filterAll,
+	children: filterAll,
+	traitChunks: filterAll,
+};
+
+const commands: CommandRegistry = [insertExample, doubleInsertExample];
+
+// Some dummy schema related data.
+const foo: Definition = 'Foo' as Definition;
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:70]: Let's replace these with globally unique definitions soon. Even just using UUIDs instead of these short strings would be fine.

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+export async function exampleApp(tree: SharedTree): Promise<void> {
+	// Perform a full checkout of all data.
+	const checkout = await tree.checkout(commands, fetchAll);
+
+	// Example app policy: this app just watches the tree, and adds bar(baz) subtree under any new `foo` node.
+	checkout.on('viewChange', (before, after) => {
+		const delta = before.delta(after);
+		for (const added of delta.added) {
+			const n: TreeNodeViewReadonly = checkout.contextualizeAnchor(anchorDataFromNodeId(added));
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:84]: contextualizeAnchor(anchorDataFromNodeId(...)) feels way too complicated. We need a model in which the parameter is [the common thing you'd need] and one can just start working with it. If there's some esoteric case we must support, let's add a separate esoteric change handler one can register (or a parameter to the registration, or something.

Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:84]: Minor comment: I think "n" is probably not a great name for a node. How about "node"?

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+			if (n.definition === foo) {
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:85]: It's unclear to me how this detects "any new foo node". Is delta.added going to include every single new node inserted, or just the roots of any inserted nodes? The former sounds expensive, and if we're going to support it I think we'd want to separate this into something like delta.treesAdded and delta.allNodesAdded.

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+function newNodeId(): StableId {
+	return v4() as StableId;
+}
+
+export const doubleInsertExample = {
+	id: '08bac27d-632f-48bb-834a-90af8d67ca60' as CommandId,
+	run: (context: CommandContext, _: {}, { parent }: { parent: TreeNode }): TreeNode => {
+		const a = context.runCommand(
+			insertExample,
+			{ definition: context.stabilize(bar), identifier: newNodeId() },
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:51]: It's unclear to me why we need to stabilize bar if we're passing it to our own API. If it's a ShortId, shouldn't it still be the same across all contexts? (Yes, we can have temporary ShortIds, but I would have thought updates to those are a behind-the-scenes detail, and in any case a client can do anything about it while offline.)

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+			{ place: parent.childrenFromTrait(testTrait).start }
+		);
+
+		return context.runCommand(
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:55]: I like the API aspects of this, but the specific example might make readers wonder why we didn't just pass a larger "builder" tree in. (I realize this probably isn't the final builder API.)

File: experimental/dds/tree/src/interface-prototype/Anchors.ts (hunk 1 above)
Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Anchors.ts:1]: Hopefully its more clear now.

What you are describing is more at the abstraction level of StablePlace in experimental\dds\tree\src\default-edits\PersistedTypes.ts . The purpose of this proposed API is to provide something nicer to work with than those types + snapshots (aka the API you get in Checkout), while capturing enough information to enable replaying Commands as higher levels of amstraction than raw shared tree edits when desired (and the command is available)

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+			if (n.definition === foo) {
+				checkout.runCommand(doubleInsertExample, {}, { parent: n as TreeNodeData });
+			}
+		}
+	});
+
+	const treeRoot: TreeNodeViewReadonly = checkout.contextualizeAnchor(root);
+	console.log(treeRoot.queryJsonSnapshot.subtree);
+	await wait(10000);
+	console.log(treeRoot.queryJsonSnapshot.subtree);
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts:99]: What are we expecting to happen here? A comment would be helpful for this block of code.

File: experimental/dds/tree/src/interface-prototype/Examples.ts [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+function wait(ms) {
+	return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+// Note: Redo is the same as undoing an undo (at this level).
+export const undo: Command<{ editId: StableId }, {}, void> = {
```
Comment (PaulKwMicrosoft) [inline experimental/dds/tree/src/interface-prototype/Examples.ts]: Not sure why we'd put this in the examples. Won't this be built into SharedTree?

File: experimental/dds/tree/src/interface-prototype/Sequence.ts [hunk 18]
```diff
@@ -0,0 +1,16 @@
+// Sequence generic
+// TODO: are sequences mutable? Are locations usable across mutations? What about Iterators?
+
+export interface Sequence<Element, Place = Element | undefined> {
+	[Symbol.iterator](): SequenceIterator<Element, Place>;
+	iteratorFromEnd(): SequenceIterator<Element, Place>;
+	areInOrder(first: Element, second: Element): boolean;
+	readonly length: number;
+	indexOf(element: Element): number;
+	elementAtIndex(index: number): Element;
```
Comment (DLehenbauer) [inline experimental/dds/tree/src/interface-prototype/Sequence.ts:10]: Consider just 'at(index) to match [TC39 propasal](https://tc39.es/proposal-relative-indexing-method/).

File: experimental/dds/tree/src/interface-prototype/Checkout.ts [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+/**
+ * Specifies what data to prefetch.
+ * Where the filters return false, Placeholders are allowed.
+ * Returning true for all with fetch everything and produce no placeholders.
+ */
+export interface PrefetchFilter {
+	value: Filter<Definition>;
+	children: Filter<Definition>;
+	traitChunks: Filter<TraitLabel>;
+}
```
Comment (DLehenbauer) [inline experimental/dds/tree/src/interface-prototype/Checkout.ts:196]: I think you'll want something declarative that you can ship to the server like, [JMESPath](https://jmespath.org/).

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 17 above)
Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Examples.ts]: Modifications to the tree are done through commands. Undo modifies the tree so it is a command.

Like other commands, it has some higher level logic that could be replayed, and thus can benefit from the hierarchal nature of commands.

Making it a command allows for a simpler abstraction at this level, and I see no downside.

If undo is going to be a command, I figured it would make a reasonable example of one: I figured that would be better than simply omitting it from this prototype. Its particularly interesting because of its history access requirements and the desire to make it synchronous to be compatible with existing users of experimental/tree.

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 15 above)
Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Examples.ts:55]: You could pass a large one in, but then either the example would not show nested commands, or would be a larger example.

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 14 above)
Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Examples.ts:51]: the extra options passed to Commands have to be able to be persisted and reapplied in another instance of the application when/of replaying the command. Since we have no reflection information for what the values in the options are, we have no way to convert what ever runtime Id representation we have into something more stable when persisting it (we can only tell that its a number, string, etc). Thus we have to use something stable directly.

I'm not happy with how this turned out: its pretty ugly. I don't see a better approach unless we just allow using the full persistable Ids more APIs (and add some more static types for them).

In realistic usage it would be less bad since you are likely to have the full Ids in scope from your schema objects, and also passing definitions to commands probably won't be common (I imagine anchors and various primitives will likely make up most of the arguments)

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 12 above)
Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Examples.ts]: Changed. I don't think `node` is great as its just redundant (and less specific) type information. I think `inserted` actually provides some information to the reader.

Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Examples.ts]: you only end up needing anchorDataFromNodeId here because we haven't wrapped up the delta and snapshot APIs in something that produces anchors. I added a better comment to anchorDataFromNodeId about this.

File: experimental/dds/tree/src/interface-prototype/Sequence.ts (hunk 18 above)
Comment (PaulKwMicrosoft) [reply experimental/dds/tree/src/interface-prototype/Sequence.ts:10]: I like that idea.

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 10 above)
Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Examples.ts]: welcome to TypedEventEmitter. Its typesafe with statically typed string constants that look unsafe, but are actually strongly typed, but also compile anyway if you get them wrong, and fails to type check the call back signature. This is how Node does its javascript eventing, and fluid uses that with some only partially typesafe static typing.

Note that this event is not new in this prototype and is the same event currently used by experimental/tree, and using string names here is how other fluid code does it, and gets better static typing that using our enum (which is better at making sure you get the correct value.).

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 9 above)
Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Examples.ts:78]: it does not. This `as` is required or this will not build.

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 8 above)
Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Examples.ts]: done

File: experimental/dds/tree/src/interface-prototype/Checkout.ts [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+	 * If receiver of the event tracks the before snapshot, this pattern might work better.
+	 */
+	runCommand<TOptions extends Serializable, TAnchorSet extends AnchorSet, TResult>(
+		command: Command<TOptions, TAnchorSet, TResult>,
+		parameters: TOptions,
+		anchors: DecontextualizedAnchorSet<TAnchorSet>
+	): { editId: EditId; result: TResult } | Aborted;
+
+	// TODO: need API for async command application (runs async commands and also supports running synchronous commands with retry after loading data)
+}
```
Comment (DLehenbauer) [inline experimental/dds/tree/src/interface-prototype/Checkout.ts:226]: I think your main concepts will be easier for a TypeScript developer to follow if you hang the primary API off of a single object.  (In this case, I think that object is 'Checkout'.)

```ts
export interface Checkout {
    // read
    query(..)

    // primitive ops
    insert(..)
    detach(..)
    
    // commands
    runCommand(..)
    
    ...etc...
}
```

I think the above pivot, where anchors are passed to insert(), detach(), etc. also removes the need for "contextualized" vs. "non-contextualized" variants?

Comment (DLehenbauer) [reply experimental/dds/tree/src/interface-prototype/Checkout.ts:226]: I think you can similarly eliminate the "view" types by moving the reader APIs of checkout to a base class or interface.

File: experimental/dds/tree/src/interface-prototype/Checkout.ts (hunk 19 above)
Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Checkout.ts:196]: Fluid blob storage does not support such things. We are relying on somewhat large logical tree chunks to keep the round trip count low.

If we do end up deploying custom server code to reduce round trip delays when prefetching, something like that could be a good idea.

That said, I think this is yet another augment for having a schema based typed API which included chunking.
Schemas which include prefetch/async requirements, are a declarative query language which work well here.

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 14 above)
Comment (PaulKwMicrosoft) [reply experimental/dds/tree/src/interface-prototype/Examples.ts:51]: I guess I still don't see why the ShortId wouldn't be stable, even if reapplied in another instance of the application replaying the command. If the command were to be applied to another *document*, that would be another matter. I suppose we *might* want to support that in some complex mutli-document simultaneous merge... but is that the scenario you had in mind or is there something simpler that I'm missing? One idea might be to put all tree-related parameters in the "anchors" block, though we might need a different name for it then.

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 9 above)
Comment (PaulKwMicrosoft) [reply experimental/dds/tree/src/interface-prototype/Examples.ts:78]: I'm not suggesting you drop the `as`, I'm suggesting you drop the `: TraitLabel`, as it can be inferred.

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 12 above)
Comment (PaulKwMicrosoft) [reply experimental/dds/tree/src/interface-prototype/Examples.ts]: Ok, something other than `node`, then. I just felt that `n` sounded like a number.

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 17 above)
Comment (PaulKwMicrosoft) [reply experimental/dds/tree/src/interface-prototype/Examples.ts]: Hm, I'm not sure an app *could* write undo, once we get through all the most complex scenarios. I don't think we should have this in the examples. Let's pick something else instead.

File: experimental/dds/tree/src/interface-prototype/Sequence.ts (hunk 18 above)
Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Sequence.ts]: done

File: experimental/dds/tree/src/interface-prototype/Examples.ts (hunk 14 above)
Comment (CraigMacomber) [reply experimental/dds/tree/src/interface-prototype/Examples.ts:51]: I believe you have some ShortID proposal somewhere thats relevant, and I assume based on your response provides stable ShortIds. Can you link it?
